lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp

```diff
@@ -414,19 +343,243 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
+
+    // Layout for the store from registers to shared memory.
+    //
+    // Note: If two threads in the same warp write to the same shmem offset, the
+    // hardware resolves that without a stall or a bank conflict.  Therefore we
+    // don't need to avoid duplicate writes.
+    LinearLayout shmemStoreLayout = srcLayout.invertAndCompose(sharedLayout);
+    const int shmemAllocatedNumElems =
+        getTotalSize<unsigned>(scratchConfig.paddedRepShape);
+    assert(shmemStoreLayout.getOutDimSize(kOffset) <= shmemAllocatedNumElems);
+
+    // Layout for the load from shmem to registers.
+    LinearLayout shmemLoadLayout = dstLayout.invertAndCompose(sharedLayout);
+
+    // Check that the `register` fully determines the `iteration`.  That is,
+    // each thread does exactly the same reads and writes to shmem on each
+    // iteration, just with different input/output registers.
+    assert(
+        shmemStoreLayout.sublayoutIsZero({kLane, kWarp, kBlock}, {kIteration}));
+    assert(
+        shmemLoadLayout.sublayoutIsZero({kLane, kWarp, kBlock}, {kIteration}));
+
+    // iteration -> registers
+    SmallVector<SmallVector<int>> inRegsForIter =
+        collectRegsForIter(ctx, shmemStoreLayout);
+    SmallVector<SmallVector<int>> outRegsForIter =
+        collectRegsForIter(ctx, shmemLoadLayout);
+
+    // Avoid emitting a call to getClusterCTAId if we know it's always 0.
+    std::optional<Value> ctaId;
+    if (conversion.getInDimSize(kBlock) > 1)
+      ctaId = targetInfo.getClusterCTAId(rewriter, loc);
+
+    // Map src/dst registers to each iteration.
+    Value smemBase =
+        LLVM::getSharedMemoryBase(loc, rewriter, op.getOperation());
+    Type elemTy = inVals[0].getType();
+    auto outSize = shmemLoadLayout.getInDimSize(kRegister);
+    auto iterations = sharedLayout.getInDimSize(kIteration);
+
+    auto getVecAddrAndBlock = [&](ArrayRef<Value> offsetAndBlock)
+        -> std::pair<Value, std::optional<Value>> {
+      std::optional<Value> block =
+          isCrossCTA ? std::make_optional(offsetAndBlock[2]) : std::nullopt;
+      Value offset = offsetAndBlock[0];
+      if (scratchConfig.paddedSize > 0) {
+        assert(llvm::isPowerOf2_32(scratchConfig.paddedStride));
+        assert(llvm::isPowerOf2_32(scratchConfig.paddedSize));
+        auto rshiftVal = llvm::Log2_32(scratchConfig.paddedStride);
+        auto lshiftVal = llvm::Log2_32(scratchConfig.paddedSize);
+        offset = add(shl(lshr(offset, i32_val(rshiftVal)), i32_val(lshiftVal)),
+                     offset);
+      }
+      auto vecAddr = gep(sharedPtrTy, elemTy, smemBase, offset);
+      vecAddr.setInbounds(true);
+      return {vecAddr, block};
+    };
+
+    // register idx -> Value
+    llvm::MapVector<int, Value> outVals;
+    for (int i = 0; i < iterations; i++) {
+      if (i != 0)
+        barrier();
+
+      auto &inRegs = inRegsForIter[i];
+      auto &outRegs = outRegsForIter[i];
+
+      // XXX: CSE applyLinearLayout.
```

Comment:
Maybe write this comment out a little more?  (Also if this is the only problem you have on AMD, it's just a few lines to fix it, probably just worth fixing properly...)

Reply:
I think this problem is from the original `ll` branch, which I actually don't understand...

Reply:
haha :)

AMD hit a problem with LLs generating slower code for emitIndices.  They solved it with this hack: https://github.com/triton-lang/triton/blob/main/lib/Conversion/TritonGPUToLLVM/Utility.cpp#L240

We were all confused why this is necessary because we expected LLVM to CSE this.  But it seemed to help them.

Ideally this hack would exist inside applyLinearLayout instead of where it is currently.  IIRC they said they would look into it but I think it may have dropped off their priorities list.

Reply:
We could pin AMD to check if there' any regression introduced by this PR. If not, we could avoid the hack here.

Reply:
I would recommend just doing it; we know it can generate bad code if we don't.  There's no fundamental difference between what we're doing here and what they were hitting in their case.